libff/algebra/scalar_multiplication/multiexp.tcc

```diff
@@ -143,17 +162,132 @@ T naive_plain_exp(typename std::vector<T>::const_iterator vec_start,
     return result;
 }
 
-/*
-  The multi-exponentiation algorithm below is a variant of the Bos-Coster algorithm
-  [Bos and Coster, "Addition chain heuristics", CRYPTO '89].
-  The implementation uses suggestions from
-  [Bernstein, Duif, Lange, Schwabe, and Yang, "High-speed high-security signatures", CHES '11].
-*/
-template<typename T, typename FieldT>
-T multi_exp_inner(typename std::vector<T>::const_iterator vec_start,
-                  typename std::vector<T>::const_iterator vec_end,
-                  typename std::vector<FieldT>::const_iterator scalar_start,
-                  typename std::vector<FieldT>::const_iterator scalar_end)
+template<typename T, typename FieldT, multi_exp_method Method,
+    typename std::enable_if<(Method == multi_exp_method_BDLO12), int>::type = 0>
+T multi_exp_inner(
+    typename std::vector<T>::const_iterator bases,
+    typename std::vector<T>::const_iterator bases_end,
+    typename std::vector<FieldT>::const_iterator exponents,
+    typename std::vector<FieldT>::const_iterator exponents_end)
+{
+    UNUSED(exponents_end);
+    size_t length = bases_end - bases;
+
+    // empirically, this seems to be a decent estimate of the optimal value of c
+    size_t log2_length = log2(length);
+    size_t c = log2_length - (log2_length / 3 - 2);
+
+    const mp_size_t exp_num_limbs =
+        std::remove_reference<decltype(*exponents)>::type::num_limbs;
+    std::vector<bigint<exp_num_limbs> > bn_exponents(length);
+    size_t num_bits = 0;
+
+    for (size_t i = 0; i < length; i++)
+    {
+        bn_exponents[i] = exponents[i].as_bigint();
+        num_bits = std::max(num_bits, bn_exponents[i].num_bits());
+    }
+
+    size_t num_groups = (num_bits + c - 1) / c;
+
+    T result;
+    bool result_nonzero = false;
+
+    for (size_t k = num_groups - 1; k <= num_groups; k--)
```

Comment:
nit: It's better convention if this reads,
```c
for (size_t k = num_groups - 1; k >= 0; k--)
```

Reply:
@howardwu That wouldn't work, because size_t is unsigned, so (k >= 0) is always trivially true. See

Reply:
Haha yes, I wasn't putting two and two together at that hour, thanks